Add TagHelper documentation resolution to TagHelperDescriptorResolver.

NTaylorMullen opened the pull request:
- Added `TagHelperUseageDescriptor` and associated factory for the `TagHelperDescriptorFactory` to utilize.
- `TagHelperUseageDescriptor`s are only created during design time.
- CoreCLR is not supported for XML documentation resolution for now. Can revisit this later when we have better tooling integration with CoreCLR.

#352

NTaylorMullen commented:
/cc @dougbu @pranavkm

NTaylorMullen commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,144 @@
+                // to the cache to not constantly look for new XML files.
+                if (xmlDocumentationFile != null)
+                {
+                    var xmlDocumentation = XDocument.Load(xmlDocumentationFile.FullName);
```
Suggestion by @rynowak is to break out the parsing code of the XML file into a stand alone class in-case we want to utilize it in other areas down the line.

NTaylorMullen commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,136 @@
+            // Members will be null if there is no associated XML file for the provided typeAssembly.
+            if (documentationProvider != null)
+            {
+                if (documentationProvider.HasDocumentation(id))
```
Each one of these method calls essentially calls through to linq to find the member and then provide information on it. Options:
1. Leave as is
2. Cache the member lookup in the `documentationProvider`
3. Provide an extra abstraction (`XmlDocumentationMember`) to make retrieval of all these bits easy (kind of like `FileInfo` but for an xml doc member).

dougbu replied:
4. Do the work _once_. As mentioned in next class, get rid of `HasDocumentation()` and add `string.IsNullOrEmpty()` checks before creating the usage descriptor.

NTaylorMullen commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,136 @@
+
+                return assemblyXmlDocumentationFile;
+            }
+            catch (PathTooLongException)
```
Instead of doing this could a custom exception thrown by the `XmlDocumentationProvider` that can be caught OR could add a property to the provider like `XmlFileResolved`. Problem with adding a property would then add the requirement to throw whenever one of the `GetSummary` or `GetRemarks` methods is invoked.

Adding this comment for discussion.

dougbu replied:
Not creating a `XmlDocumentationProvider` here, just evaluating paths. Leave as-is.

NTaylorMullen commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,136 @@
+            return null;
+        }
+
+        private static FileInfo GetXmlDocumentationFile(string typeAssemblyLocation)
```
Debated moving the functionality of this method into another type such as `XmlDocumentationLocator` but thought it might be overkill for now.

NTaylorMullen commented:
Updated by splitting out the `XmlDocumentationProvider` and added several comments to the PR for discussion.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -44,17 +44,27 @@ public static class TagHelperDescriptorFactory
+#if !DNXCORE50
+            // The TagHelperUseageDescriptor is consumed at design time to provide IntelliSense when typing HTML
+            // elements or attributes that are tag helper related.
+            typeUseageDescriptor = designTime ? TagHelperUseageDescriptorFactory.CreateDescriptor(typeInfo) : null;
```
typo throughout this PR; should be `Usage`

NTaylorMullen replied:
:facepalm:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -246,7 +269,8 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
                         defaultPrefix: descriptor.Name + "-",
-                        isInvalid: out isInvalid);
+                        isInvalid: out isInvalid,
+                        designTime: designTime);
```
leave `out` parameters last. less important but would also be good to place `designTime` consistently just before (or after) `errorSink`.

NTaylorMullen replied:
named parameters failed me. That's how it is in the method signature :/

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -414,11 +441,35 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
+            return new TagHelperAttributeDescriptor(
+                attributeName,
+                property.Name,
+                property.PropertyType.FullName,
```
Change would mess up `isIndexer` cases; pass `typeName`.

NTaylorMullen replied:
:+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -121,6 +125,7 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
         protected virtual IEnumerable<TagHelperDescriptor> ResolveDescriptorsInAssembly(
             string assemblyName,
             SourceLocation documentLocation,
+            bool designTime,
```
Remove this parameter. Any subclass must provide value when calling a `TagHelperDescriptorResolver` constructor. So that subclass can have its own `private bool _designTime` field.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/project.json:
```diff
@@ -6,11 +6,21 @@
         "Microsoft.Framework.ClosedGenericMatcher.Sources": { "type": "build", "version": "1.0.0-*" },
         "Microsoft.Framework.CopyOnWriteDictionary.Sources": { "type": "build", "version": "1.0.0-*" },
-        "Microsoft.Framework.NotNullAttribute.Sources": { "type": "build", "version": "1.0.0-*" }
+        "Microsoft.Framework.NotNullAttribute.Sources": { "type": "build", "version": "1.0.0-*" },
```
invalid JSON

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperAttributeDescriptor.cs:
```diff
@@ -24,6 +24,42 @@ internal TagHelperAttributeDescriptor([NotNull] string name, [NotNull] PropertyI
+                  propertyName,
+                  typeName,
+                  isIndexer,
+                  isStringProperty: string.Equals(typeName, typeof(string).FullName, StringComparison.Ordinal),
```
pass `isStringProperty`

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperAttributeDescriptor.cs:
```diff
@@ -46,13 +82,15 @@ internal TagHelperAttributeDescriptor([NotNull] string name, [NotNull] PropertyI
             [NotNull] string typeName,
-            bool isIndexer)
+            bool isIndexer,
+            TagHelperUseageDescriptor useage)
```
Parameter and property need longer names. suggest `usageDescriptor`. Same in `TagHelperDescriptor`.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperAttributeDescriptor.cs:
```diff
@@ -24,6 +24,42 @@ internal TagHelperAttributeDescriptor([NotNull] string name, [NotNull] PropertyI
+                  isStringProperty: string.Equals(typeName, typeof(string).FullName, StringComparison.Ordinal),
+                  useage: null)
+        {
+        }
```
In real PR, update tests to use more-complete constructors and remove these. Same in `TagHelperDescriptor`.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if !DNXCORE50 // Not used in coreclr
```
Remove useless comment or describe what doesn't work (why this class isn't used with Core).

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            var associatedMemeber = _members
+                .FirstOrDefault(element =>
+                    string.Equals(element.Attribute("name").Value, id, StringComparison.Ordinal));
```
Is `name` attribute required? If not need `?.`.

NTaylorMullen replied:
Yes it's required.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,66 @@
+
+        public bool HasDocumentation(string id)
+        {
+            return GetMember(id) != null;
```
`<member>` may not contain `<summary>` or `<remarks>`. Either check that one or both exist or change method name to something like `MayHaveDocumentation()`. But best option is probably to remove this method and change caller to confirm at least one of `summary` or `remarks` are not `null` or empty before creating a usage descriptor.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,136 @@
+    public static class TagHelperUseageDescriptorFactory
+    {
+        private static readonly Dictionary<string, XmlDocumentationProvider> XmlDocumentationProviderCache =
+            new Dictionary<string, XmlDocumentationProvider>(StringComparer.Ordinal);
```
Have you tested and confirmed this cache is necessary? Given the short flurry of tag helper descriptor resolution compared to the lifetime of the host process, could tie up a lot of memory for little reason. Also much better to add caches once we've measured and determined exactly what needs to be cached.

For now, even though we know the count of tag helper classes per tag helper assembly is likely greater than one, do not add this.

NTaylorMullen replied:
This prevents Razor parsing from going to the file system. We've always pre-emptively added caches for file system like behaviors.

dougbu replied:
There already is a cache of `TagHelperDescriptors` so this is redundant.

NTaylorMullen replied:
That cache isn't in Razor, it's in VS outside of our control entirely.

NTaylorMullen replied:
Don't feel strongly enough about it. Will remove.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs:
```diff
@@ -0,0 +1,136 @@
+        public static TagHelperUseageDescriptor CreateDescriptor([NotNull] PropertyInfo propertyInfo)
+        {
+            var id = XmlDocumentationProvider.GetId(propertyInfo);
+            var declaringTypeAssembly = propertyInfo.DeclaringType.GetTypeInfo().Assembly;
```
Throughout this PR, remove `Type` from `TypeAssembly` names. Every assembly contains types. `declaringAssembly` here, `assemblyLocation` (or just `location`) a few lines down, and so on.

dougbu commented:
:watch:

NTaylorMullen commented:
Updated.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,129 @@
+            var xmlDocumentationFile = GetXmlDocumentationFile(assemblyLocation);
+
+            // We only want to process the file if it exists. In the case it doesn't, a null value will be added
+            // to the cache to not constantly look for new XML files.
```
Comment is no longer correct

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,129 @@
+                documentationProvider = new XmlDocumentationProvider(xmlDocumentationFile.FullName);
+            }
+
+            // Members will be null if there is no associated XML file for the provided assembly.
```
"Members" ??

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,129 @@
+            }
+
+            // Members will be null if there is no associated XML file for the provided assembly.
+            if (documentationProvider != null)
```
Can be combined with the above `if` and can move `var documentationProvider` inside the block.

dougbu commented:
Design outline looks good. Ready for the real thing.

NTaylorMullen commented:
Updated, this is no longer in Design. /cc @dougbu

NTaylorMullen commented:
Note: Was unable to validate a full end-to-end `TagHelperDescriptorFactory.CreateDescriptor`s creating `TagHelperUsageDescriptor`s due to the lack of XML files from our runtime (it loads projects into memory) and the requirement to have full reflection capability on `Type`s (if we wanted to attempt to fake a `Type`).

There was one alternative I thought of: Create a pre-build script that generates a package and then forces the test project to reference the package instead of the project on test run. Seemed to heavy handed for a minimal scenario so decided not to.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -38,23 +38,37 @@ public static class TagHelperDescriptorFactory
+
+#if !DNXCORE50
+            // The TagHelperUsageDescriptor is consumed at design time to provide IntelliSense when typing HTML
+            // elements or attributes that are tag helper related.
```
Don't see much value in this comment or its repetition later in this file. Reads like `<remarks>` in the descriptor classes and says nothing about the line of code. Suggest removing or perhaps just saying `// TagHelperUsageDescriptor information is not needed at runtime because it used only for IntelliSense.`

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -26,14 +26,16 @@ public class TagHelperDescriptorResolver : ITagHelperDescriptorResolver
-        public TagHelperDescriptorResolver()
-            : this(new TagHelperTypeResolver())
+        /// <param name="designTime">Indicates if the <see cref="TagHelperDescriptorResolver"/> should run in
+        /// design time mode</param>
```
Be more specific, similar to `TagHelperDescriptorFactory` doc comments e.g. `Indicates whether resolved <see cref="TagHelperDescriptor"/>s should include design time-specific information`.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    /// <summary>
+    /// Factory for <see cref="TagHelperUsageDescriptor"/>s from <see cref="Type"/>s and <see cref="PropertyInfo"/>s.
```
Missing a word like "extracting" or "providing" after "for".

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+                assemblyLocation = Uri.UnescapeDataString(uri.Path);
+
+                // Still couldn't resolve a valid assemblyLocation.
+                if (string.IsNullOrWhiteSpace(assemblyLocation))
```
Move this check after the containing `if`. Won't currently be reached if `assemblyLocation` is `null` or whitespace and `CodeBase` is also `null` or whitespace.

NTaylorMullen replied:
It's handled below but it does still attempt to locate the XML file. Will move down :+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            var assemblyLocation = assembly.Location;
+
+            if (string.IsNullOrWhiteSpace(assemblyLocation) && !string.IsNullOrWhiteSpace(assembly.CodeBase))
```
Why `IsNullOrWhiteSpace()` and not `IsNullOrEmpty()`?

NTaylorMullen replied:
No reason, being over protective. Will change.

Eilon commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Fix filename casing.

NTaylorMullen replied:
Gah, fixed it. Git doesn't like me.

Eilon commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+
+                return assemblyXmlDocumentationFile;
+            }
+            catch (PathTooLongException)
```
:frowning:

Eilon replied:
But seriously, what's so special about PathTooLong? How was the original assembly loaded to begin with if the path is too long?

NTaylorMullen replied:
:+1:, I added it because it was the only other exception the `Path` utilities could throw (I already handle the invalid path characters). Will remove.

Eilon commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+            var culture = CultureInfo.CurrentCulture;
+
+            // Following the fall-back process defined by:
+            // https://msdn.microsoft.com/en-us/library/sb6a8618.aspx#cpconpackagingdeployingresourcesanchor1
```
Did you determine if this is correct for VS language settings?

dougbu replied:
This class assumes XML files will always be named identically to the corresponding satellite assembly and found in the same directories. The referenced page doesn't describe how XML files are found, just those satellites. And it's about desktop applications, not (DNX-created) localized packages.

NTaylorMullen replied:
@Eilon talked with @ToddGrun about how the refs are picked up to provide localized text for other pieces in the editor (HTML editor etc.) and he said that it's just handled via `CultureInfo.CurrentCulture`, they don't do anything special to consume the VS specific setting.

@dougbu This isn't for DNX created packages: today DNX does not support satellite assemblies; @davidfowl said it's currently on the todo. When they do light up support for localized bits we'll revisit this and add support if it's needed (hopefully just works).

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+
+            var xmlDocumentationFile = GetXmlDocumentationFile(assemblyLocation);
+
+            // We only want to process the file if it exists.
```
In general try to avoid "we" and just describe what's going on. Here `// Only process the file if it exists.`

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+        private static FileInfo GetXmlDocumentationFile(string assemblyLocation)
+        {
+            if (string.IsNullOrWhiteSpace(assemblyLocation) ||
+                assemblyLocation.IndexOfAny(InvalidPathCharacters) != -1)
```
Directly using the file system. Why start with checks the file system does?

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+                    Path.Combine(assemblyDirectory, assemblyXmlDocumentationName));
+
+                // If there's not an XML file side-by-side the .dll it may exist in a culture specific directory.
+                if (!assemblyXmlDocumentationFile.Exists)
```
If `assemblyXmlDocumentationFile.Exists`, end up ignoring the culture. Suggest instead falling back to `assemblyDirectory` if the `Select()` below fails to find a culture-specific file.

NTaylorMullen replied:
:+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+                var assemblyDirectory = Path.GetDirectoryName(assemblyLocation);
+                var assemblyName = Path.GetFileName(assemblyLocation);
+                var assemblyXmlDocumentationName = Path.ChangeExtension(assemblyName, ".xml");
+                var culture = CultureInfo.CurrentCulture;
```
Unused variable.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+                {
+                    var fallbackDirectories = GetCultureFallbackDirectories();
+                    assemblyXmlDocumentationFile = fallbackDirectories
+                        .Select(fallbackDiretory =>
```
typo s/b `fallbackDirectory`

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,100 @@
+namespace Microsoft.AspNet.Razor.Runtime
+{
+    /// <summary>
+    /// Enables retrieval of summary and remarks XML documentation from an XML documentation file.
```
Huh, what is this "enables" about? Suggest `Retrieves ...` or `Extracts ...`

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/XMLDocumentationProvider.cs:
```diff
@@ -0,0 +1,100 @@
+            return associatedMemeber;
+        }
+
+        private static string GetUnnestedFullName(Type type)
```
Why is this useful? Tag helper classes cannot be nested.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperAttributeDescriptorComparer.cs:
```diff
@@ -31,7 +31,8 @@ protected TagHelperAttributeDescriptorComparer()
         /// <see cref="TagHelperAttributeDescriptor.Name"/>, <see cref="TagHelperAttributeDescriptor.PropertyName"/>,
-        /// and <see cref="TagHelperAttributeDescriptor.TypeName"/>. Ignores
+        /// <see cref="TagHelperAttributeDescriptor.TypeName"/>, and
+        /// <see cref="TagHelperAttributeDescriptor.UsageDescriptor"/>. Ignores
```
All that needs to change in this file is the comments. Add `UsageDescriptor` to the list of ignored properties because it (like `IsStringProperty`) can be inferred directly from `TypeName`.

NTaylorMullen replied:
That assumes that both `TagHelperAttributeDescriptor`s come from the same assembly which is wrong (we don't have the assembly on the descriptor).

dougbu replied:
My comment assumes we never compare attribute descriptors except within tag helper descriptors. Where is that invalidated? If it's incorrect, a lot more code needs to change.

NTaylorMullen replied:
We don't, actually after doing a find for this attribute descriptor comparer it's only used in tests. Going to move it there in the form of `CaseSensitiveTagHelperAttributeDescriptorComparer`. Will do the same for `TagHelperUsageDescriptorComparer`.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperUsageDescriptor.cs:
```diff
@@ -0,0 +1,32 @@
+namespace Microsoft.AspNet.Razor.TagHelpers
+{
+    /// <summary>
+    /// A metadata class describing how to use a tag helper at design time.
```
The part about `at design time` has nothing to do with this class; that choice is elsewhere. Suggest `A metadata class containing information about tag helper use.`

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperUsageDescriptor.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        /// <summary>
+        /// A summary on how to use a tag helper.
```
on -> of

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperUsageDescriptor.cs:
```diff
@@ -0,0 +1,32 @@
+        public string Summary { get; }
+
+        /// <summary>
+        /// Remarks on how to use a tag helper.
```
on -> about

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperUsageDescriptorComparer.cs:
```diff
@@ -0,0 +1,61 @@
+    /// An <see cref="IEqualityComparer{TagHelperUsageDescriptor}"/> used to check equality between
+    /// two <see cref="TagHelperUsageDescriptor"/>s.
+    /// </summary>
+    public class TagHelperUsageDescriptorComparer : IEqualityComparer<TagHelperUsageDescriptor>
```
Class either doesn't need to exist or should be a test class.

NTaylorMullen replied:
https://github.com/aspnet/Razor/pull/419#discussion_r32958698

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorComparer.cs:
```diff
@@ -31,7 +31,8 @@ protected TagHelperDescriptorComparer()
         /// <see cref="TagHelperDescriptor.AssemblyName"/>, <see cref="TagHelperDescriptor.TagName"/>,
-        /// and <see cref="TagHelperDescriptor.RequiredAttributes"/>.
+        /// <see cref="TagHelperDescriptor.RequiredAttributes"/>, and
+        /// <see cref="TagHelperDescriptor.UsageDescriptor"/>.
```
This class should not change. `UsageDescriptor` can be inferred directly from `TypeName`.

NTaylorMullen replied:
Unlike the `TagHelperAttributeDescriptorComparer` :+1: to this. It can be inferred from both `TypeName` and `AssemblyName`.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperAttributeDescriptor.cs:
```diff
@@ -39,6 +40,8 @@ internal TagHelperAttributeDescriptor([NotNull] string name, [NotNull] PropertyI
         /// Otherwise this <see cref="TagHelperAttributeDescriptor"/> is used for property assignment.
         /// </param>
+        /// <param name="usageDescriptor">The <see cref="TagHelperUsageDescriptor"/> that indicates how to use the
+        /// tag helper attribute during design time.</param>
```
No indications here and "during design time" is not worth mentioning. For all documentation of `UsageDescriptor` parameters and properties, say `... that contains information about use of this attribute.` or `... this tag helper`. If you want to add `<remarks>` mentioning the properties are `null` at runtime, I guess that's OK but that ignores all the other cases in which the descriptor will be `null`.

NTaylorMullen replied:
Wont add remarks but will change wording to do `contains information about use of this attribute` :+1:

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperUsageDescriptorComparer.cs:
```diff
@@ -0,0 +1,61 @@
+            if (descriptor == null)
+            {
+                return 0;
+            }
```
Remove this and add `[NotNull]` to the parameter. Method is _supposed_ to `throw` if passed a `null` value.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/CommonTagHelpers.cs:
```diff
@@ -32,4 +35,36 @@ public class NonPublicAccessorTagHelper : TagHelper
+    public class DocumentedTagHelper : TagHelper
+    {
+        public static readonly string XmlTestFileLocation =
+            Directory.GetCurrentDirectory() + "/TestFiles/TagHelperDocumentation.xml";
```
This doesn't belong here.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/project.json:
```diff
@@ -3,6 +3,7 @@
     "dependencies": {
         "Microsoft.AspNet.Razor.Runtime": "4.0.0-*",
         "Microsoft.AspNet.Testing": "1.0.0-*",
+        "Moq": "4.2.1312.1622",
```
Bad practise to put a runtime-specific dependency in the general list. Move under frameworks / dnx451 / dependencies.

NTaylorMullen replied:
:+1:

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,316 @@
+            {
+                // Faking localized dlls only to force the TagHelperUsageDescriptorFactory to search for localized XML.
+                var defaultENGBLocation = DocumentedTestAssemblyLocation.Replace(".dll", "-en-GB.dll");
+                var defaultENLocation = DocumentedTestAssemblyLocation.Replace(".dll", "-en.dll");
```
Something is wrong if tests using these locations pass. Product code uses locale-specific sub-directories, not locale-specific XML file names.

NTaylorMullen replied:
We were spoofing the names before to avoid a hit on the XML file that's located in the non-localized folders. Now that we're checking localized folders first prior to an XML file next to the DLL i'll adjust the tests accordingly.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,141 @@
+                    assemblyXmlDocumentationFile = fallbackDirectories
+                        .Select(fallbackDiretory =>
+                            new FileInfo(
+                                Path.Combine(assemblyDirectory, fallbackDiretory, assemblyXmlDocumentationName)))
```
A partial implementation of the process documented at https://msdn.microsoft.com/en-us/library/sb6a8618.aspx#cpconpackagingdeployingresourcesanchor1  Does not handle fallback assembly names or the precedence of a locale-specific filename in the same directory as the main assembly over one found in a sub-directory. Also does not handle anything about assemblies that are in the GAC, where XML files are likely found only in Reference Assemblies directories.

Eilon replied:
No need to support GAC reference for tag helpers, because they will never be in the GAC.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,316 @@
+                    {
+                        CreateDocumentationTagHelperTestType(defaultLocation, codeBase: null),
+                        new TagHelperUsageDescriptor(
+                            summary: "The summary for the type.",
```
`<summary>` and `<remarks>` rarely contain just simple text. Need to test more common content such as `<see cref="T:Microsoft.AspNet.Razor.Runtime.TagHelpers.ITagHelper"/> implementation targeting &lt;a&gt; elements.` and `Must be <c>null</c> if <see cref="P:Microsoft.AspNet.Mvc.TagHelpers.AnchorTagHelper.Route"/> is non-<c>null</c>.`.

NTaylorMullen replied:
:+1: found an issue with `.Value` on `XElement`s (it doesn't stringify contents that are also tags). Adding several use-cases of this all over.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/XmlDocumentationProviderTest.cs:
```diff
@@ -0,0 +1,173 @@
+        public void CanReadXml()
+        {
+            // Act. Ensuring that reading the Xml file doesn't throw.
+            var xmlDocumentationProvider = new XmlDocumentationProvider(DocumentedTagHelper.XmlTestFileLocation);
```
Remove `var xmlDocumentationProvider = `

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/XmlDocumentationProviderTest.cs:
```diff
@@ -0,0 +1,173 @@
+                {
+                    {
+                        XmlDocumentationProvider.GetId(DocumentedTagHelperTypeInfo),
+                        "The summary for the type."
```
Rinse and repeat comment in `TagHelperUsageDescriptorFactoryTest` about common `<summary>` and `<remarks>` content.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/XmlDocumentationProviderTest.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        [Fact]
+        public void GetId_TypeInfo_NestedClass()
```
Remove all nested class testing. Not a real tag helper scenario.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,316 @@
+        [ReplaceCulture]
+        [MemberData(nameof(CreateDescriptor_TypeData))]
+        [MemberData(nameof(CreateDescriptor_FallbackTypeData))]
+        public void CreateDescriptor_Type(Type tagHelperType, TagHelperUsageDescriptor expectedUsageDescriptor)
```
Change all tests in the two new classes to have names that describe expectations. This one might be `CreateDescriptor_WithType_ReturnsExpectedDescriptors()`.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/XmlDocumentationProviderTest.cs:
```diff
@@ -0,0 +1,173 @@
+                return new TheoryData<string, string>
+                {
+                    {
+                        XmlDocumentationProvider.GetId(DocumentedTagHelperTypeInfo),
```
Strange to use tested class in `// Arrange`. Suggest providing the identifiers as literal strings.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/XmlDocumentationProviderTest.cs:
```diff
@@ -0,0 +1,173 @@
+        {
+            // Arrange
+            var expectedId = $"P:{typeof(DocumentedTagHelper).FullName}." +
+                nameof(DocumentedTagHelper.RemarksAndSummaryProperty);
```
Don't mix string interpolation and concatenation in the same expression. Pick one.

dougbu commented:
:watch:

NTaylorMullen commented:
Updated. Pretty Massive, things changed:

- Added support for culture specific assemblies when searching for localized XML.
- Added support for `NeutralResourcesLanguageAttribute` to allow an ultimate fall-back location when searching for XML.
- Moved `TagHelperAttributeDescriptorComparer` from the src projects into test (wasn't being used in src). This also involved changing the `CaseSensitiveTagHelperAttributeDescriptorComparer` to now be the normal `TagHelperAttributeDescriptorComparer`.
- Added several new localization use-cases for XML documentaiton consumption.
- Updated XML files to showcase more complex scenarios.

dougbu commented:
> Added support for `NeutralResourcesLanguageAttribute` to allow an ultimate fall-back location when searching for XML.

What is this about? `[NeutralResourcesLanguage]` describes resource locations within a set of assemblies, not a location on disc.

I _could_ see a slight enhancement in the future where Razor would use the `[NeutralResourcesLanguage]`'s `CultureName` property to determine the ultimate fallback language. But do that separately -- it only complicates the initial implementation and is clearly separable.

NTaylorMullen commented:
@dougbu Unless I'm reading the documentation wrong, when `NeutralResourceLanguage.Location` is set to `UltimateResourceFallbackLocation.Satellite` it indicates to utilize the provided culture to look for the culture specific satellite assembly with xml doc.

dougbu commented:
`UltimateResourceFallbackLocation.Satellite` means the _resources_ for the neutral language (aka the assembly's final fallback locale) are found in an satellite -- and not the main assembly. `UltimateResourceFallbackLocation.CultureName` is the name of that neutral language / fallback locale.

Where does the documentation mention XML docs?

NTaylorMullen commented:
It doesn't mention XML docs specifically anywhere in the documentation. Taking the same approach as resources ensures that we can load `TagHelper` documentation in the same locale they're seeing their errors/images/etc. in. A consistent experience.

dougbu commented:
> a consistent experience

Huh? Finding things inside an assembly group is completely different from finding other files that (usually) have the same base name as the assembly or a satellite. For one thing the runtime location of the assembly group is often _not_ where XML doc files are found.

NTaylorMullen commented:
It's a final fall-back location to search. In the case that there's no XML file for the current culture, in the current directory, of various names, *then* we look for the `NeutralLangaugeAttribute`. 

Are you suggesting we don't provide any `TagHelper` documentation in that case?

NTaylorMullen commented:
Updated.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -41,9 +43,12 @@ public TagHelperDescriptorResolver()
         /// <param name="typeResolver">The <see cref="TagHelperTypeResolver"/>.</param>
-        public TagHelperDescriptorResolver(TagHelperTypeResolver typeResolver)
+        /// <param name="designTime">Indicates if the <see cref="TagHelperDescriptorResolver"/> should run in
+        /// design time mode</param>
```
Use same description as in previous constructor.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,171 @@
+            }
+
+            // Couldn't resolve a valid assemblyLocation.
+            if (string.IsNullOrWhiteSpace(assemblyLocation))
```
`IsNullOrEmpty()`

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,171 @@
+
+                if (localizedXmlFile == null)
+                {
+                    // Check for an uncultured XML file next to the assembly
```
uncultured -> "culture-neutral" :smile:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,171 @@
+                    }
+
+                    return xmlFile;
+                }).FirstOrDefault(file => file.Exists);
```
Looks like an unrolled loop a compiler has spit out in IL. Would be easier to understood if the `ExpandPaths()` method expanded all the paths and the Linq was just `Select(path => new FileInfo(path)).FirstOrDefault(file => file.Exists)`. Also avoids the extra `Exists` check and makes the linked page more relevant within `ExpandPaths()`. The method might be
``` c#
private static IEnumerable<string> ExpandPaths(
    CultureInfo culture,
    string assemblyDirectory,
    string assemblyXmlDocumentationName)
{
    do
    {
        var cultureName = culture.Name;
        var cultureSpecificFileName =
            Path.ChangeExtension(assemblyXmlDocumentationName, cultureName + ".xml");

        yield return Path.Combine(assemblyDirectory, cultureSpecificFileName);
        yield return Path.Combine(assemblyDirectory, cultureName, assemblyXmlDocumentationName));
        yield return Path.Combine(assemblyDirectory, cultureName, cultureSpecificFileName));

        culture = culture.Parent;
    } while (culture != null && culture != CultureInfo.InvariantCulture);
}
```

NTaylorMullen replied:
Hmm, I like it :+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,171 @@
+                    CultureInfo.CurrentCulture,
+                    assemblyDirectory,
+                    assemblyXmlDocumentationName);
+                FileInfo assemblyXmlDocumentationFile;
```
Confusing to have two variables of the same type here; just need one `xmlDocumentationFile` variable.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,171 @@
+                    assemblyXmlDocumentationFile = localizedXmlFile;
+                }
+
+                return assemblyXmlDocumentationFile.Exists == true ? assemblyXmlDocumentationFile : null;
```
Minor but this `Exists` check is unnecessary when using a localized file. Realize now you could do the opposite check of what used to be in the `if` block. Then just `return xmlDocumentationFile;` here. I.e.
``` c#
if (xmlDocumentationFile == null)
{
    xmlDocumentationFile = new FileInfo(
        Path.Combine(assemblyDirectory, assemblyXmlDocumentationName));
    if (!xmlDocumentationFile.Exists)
    {
        xmlDocumentationFile = null;
    }
}
```

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -136,6 +141,11 @@ public class TagHelperDescriptor
+        public IReadOnlyList<string> RequiredAttributes { get; }
+
+        /// <summary>
+        /// The <see cref="TagHelperUsageDescriptor"/> that indicates how to use the tag helper at design time.
```
Stick with the "contains information about" wording for all `TagHelperUsageDescriptor` properties and parameters.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperAttributeDescriptorComparer.cs:
```diff
@@ -0,0 +1,50 @@
+                return true;
+            }
+
+            return descriptorX != null &&
```
`descriptorY` may also be `null`; all we know is they aren't both `null`.

dougbu replied:
thx

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperAttributeDescriptorComparer.cs:
```diff
@@ -0,0 +1,50 @@
+        public int GetHashCode(TagHelperAttributeDescriptor descriptor)
+        {
+            // Ignore IsStringProperty because it is directly inferred from TypeName and thus won't vary the hash
+            // bucket. Base comparer does not care about Name case in its hash code but in tests we do.
```
Correct out-of-date comment.

dougbu commented on test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperUsageDescriptorComparer.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        public static readonly TagHelperUsageDescriptorComparer Default = new TagHelperUsageDescriptorComparer();
+
+        protected TagHelperUsageDescriptorComparer()
```
`private`

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorComparer.cs:
```diff
@@ -0,0 +1,41 @@
+    {
+        public static readonly TagHelperUsageDescriptorComparer Default = new TagHelperUsageDescriptorComparer();
+
+        protected TagHelperUsageDescriptorComparer()
```
`private`

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,385 @@
+            "Collections.Generic.IDictionary`2\" />.";
+        private const string TestRemarksAndSummaryPropertyRemarksSuffix = "<see cref=\"P:Microsoft.AspNet.Razor." +
+            "Runtime.TagHelpers.DocumentedTagHelper.SummaryProperty\" /><see cref=\"P:Microsoft.AspNet.Razor.Runtime" +
+            ".TagHelpers.DocumentedTagHelper.RemarksProperty\" />";
```
"Test" is noise, "Suffix" is misleading, and word order is confusing. So `TypeSummary`, `TypeRemarks`, `PropertySummary` (`PropertyWithSummary_Summary` seems a tad redundant), `PropertyRemarks`, `PropertyWithSummaryAndRemarks_Summary`, `PropertyWithSummaryAndRemarks_Remarks`

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,385 @@
+        public static readonly string LocalizedDocumentedTestAssemblyLocation =
+            Directory.GetCurrentDirectory() + "/TestFiles/Localized/TagHelperDocumentation.dll";
+        public static readonly string DocumentedTestAssemblyCodeBaseLocation =
+            "file:///" + DocumentedTestAssemblyLocation;
```
Better but remove "Test" here and just `DocumentedAssemblyCodeBase` for the last field.

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,385 @@
+            return CreateTestType<DocumentedTagHelper>(location, codeBase);
+        }
+
+        private static Type CreateTestType<TWrappedType>(string location, string codeBase)
```
Suggest `CreateType()`

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperUsageDescriptorFactoryTest.cs:
```diff
@@ -0,0 +1,385 @@
+            Assert.Equal(expectedUsageDescriptor, usageDescriptor, TagHelperUsageDescriptorComparer.Default);
+        }
+
+        private static Type CreateDocumentationTagHelperTestType(string location, string codeBase)
```
Suggest `CreateDocumentedTagHelperType()`

dougbu commented:
:watch:

NTaylorMullen commented:
Updated.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs:
```diff
@@ -0,0 +1,155 @@
+
+                yield return Path.Combine(assemblyDirectory, cultureSpecificFileName);
+                yield return Path.Combine(assemblyDirectory, cultureName, assemblyXmlDocumentationName);
+                yield return Path.Combine(assemblyDirectory, cultureName, cultureSpecificFileName);
```
I left comments out of my sample code. Please restore the link to `https://msdn.microsoft.com/en-us/library/sb6a8618.aspx#cpconpackagingdeployingresourcesanchor1` (above the `do`) and the short descriptions of each path e.g. `// A culture-specific XML file next to the assembly.` (with blank lines above each comment).

dougbu commented:
:shipit:
